### PR TITLE
Fix an issue caused by mismatched targetlist of Motion node

### PR DESCRIPTION
**Pull request by pengzhout**
A Motion node may has a subplan whose targetlist need to be expanded
to make sure all entries of the subplan's hashExpr are in its
targetlist, the difference of targetlist will cause Motion node parse the
tuple (especially for MemTuple) incorrectly with a mismatched tuple
description, make wrong result or even crash.

For a subplan, if we can get a Expr for distkey column from the
targetlist, the Expr must be in the targetlist or reference the
Vars in the targetlist, The old logic is, if the Expr is valid,
the Expr that reference Vars is also added to the targetlist.
For example, a subplan has targetlist (c1) and the locus whose
distkey is (c1::float8), so the Expr returned will be one
referenced Vars in the targetlist and then the old targetlist
will be extended to (c1, c1::float8). Obviously, the extension
of targetlist here is unnecessary, the Expr can evaluate from
Vars, meanwhile, motion need to transfer more columns in a tuple.

Co-authored-by: Adam Lee <ali@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/subselect_gp.out [hunk 1]
```diff
@@ -1820,3 +1820,26 @@ select * from init_main_plan_parallel where exists (select * from pg_class);
...
 
+-- A subplan whose targetlist might be expanded to make sure all entries of its
+-- hashExpr are in its targetlist, test the motion node above it also updated
+-- its targetlist, otherwise, a wrong answer or a crash happens.
+DROP TABLE IF EXISTS TEST_IN;
+NOTICE:  table "test_in" does not exist, skipping
+CREATE TABLE TEST_IN(
+    C01  FLOAT,
+    C02  NUMERIC(10,0)
+)WITH(APPENDONLY=TRUE,COMPRESSTYPE=ZLIB,COMPRESSLEVEL=5)DISTRIBUTED RANDOMLY;
```
Comment (hlinnaka) [inline src/test/regress/expected/subselect_gp.out]: Can we leave out the WITH options? The test exercises the problem regardless of what type of a table it is, right?

**Comment by hlinnaka** [comment]
Could we reorder the code in that function so that we'd add the hashExprs to the subplan's target list first, and build the Motion node only after that? That would seem cleaner than changing the Motion node's target list afterwards.

File: src/test/regress/expected/subselect_gp.out (hunk 1 above)
Comment (pengzhout) [reply src/test/regress/expected/subselect_gp.out]: Yes, I think so, will fix it, thanks.

**Comment by pengzhout** [comment]
> Could we reorder the code in that function so that we'd add the hashExprs to the subplan's target list first, and build the Motion node only after that? That would seem cleaner than changing the Motion node's target list afterwards.

Make sense.

**Comment by pengzhout** [comment]
> Could we reorder the code in that function so that we'd add the hashExprs to the subplan's target list first, and build the Motion node only after that? That would seem cleaner than changing the Motion node's target list afterwards.

@hlinnaka @guofengrichard  After a second thought, the idea of putting hashExprs to the subplan's target list seems not necessary to me, I removed the code, please help take another look.

**Comment by adam8157** [comment]
Force pushing to trigger the PR pipeline...

**Comment by danielgustafsson** [comment]
Pushed this to master after review, as the fix is needed in the next minor release of 5.X

**Pull request closed without merging**